### PR TITLE
Add Auto TLS support via CertificateSigningRequest sent to k8s API

### LINKED ISSUE #21: Add possibility for TLS automatic creation and signing externally via k8s
Eg.
https://kubernetes.io/docs/tasks/tls/managing-tls-in-a-cluster/#create-a-certificate-signing-request-object-to-send-to-the-kubernetes-api

### PR DESCRIPTION
**Pull request by nitisht**
Fixes #21

### CODE REVIEW

**Comment by nitisht** [comment]
Ready for review, need to update docs

File: pkg/controller/cluster/csr.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+)
+
+const (
+	csrType           = "CERTIFICATE REQUEST"
+	rsaPrivateKeyType = "RSA PRIVATE KEY"
+)
+
+// newPrivateKey returns randomly generated RSA private key.
+func newPrivateKey() (*rsa.PrivateKey, error) {
+	return rsa.GenerateKey(rand.Reader, constants.DefaultRSAKeySize)
```
Comment (aead) [inline pkg/controller/cluster/csr.go]: We should use ECC (ECDSA) private keys. Reasoning:
 - Some legacy systems only support modern symmetric ciphers (AES-GCM, ChaCha20-Poly1305) with ECDSA but not with RSA.
 - Smaller Key/Signature Size -> Less traffic.
 - Constant time implementation.

File: pkg/controller/cluster/csr.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	var err error
+
+	glog.V(0).Infof("Generating private key")
+	privateKey, err = newPrivateKey()
+	if err != nil {
+		glog.Errorf("Unexpected error during the RSA Key generation: %v", err)
+		return nil, nil, err
+	}
+
+	privKeyBytes := x509.MarshalPKCS1PrivateKey(privateKey)
```
Comment (aead) [inline pkg/controller/cluster/csr.go]: For ECC we need PKCS8

File: pkg/controller/cluster/csr.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: csrName,
+		},
+		Spec: certificates.CertificateSigningRequestSpec{
+			Request: encodedBytes,
+			Groups:  []string{"system:authenticated"},
+			Usages: []certificates.KeyUsage{
+				certificates.UsageDigitalSignature,
+				certificates.UsageKeyEncipherment,
```
Comment (aead) [inline pkg/controller/cluster/csr.go]: Remove this entry - We don't encrypt session keys.

File: examples/minioinstance-with-external-service.yaml [hunk 4]
```diff
@@ -30,24 +37,37 @@ spec:
...
+    #   value: "EC:2"
   ## Configure resource requests and limits for MinIO containers
   resources:
     requests:
-      memory: 256Mi
+      memory: 512Mi
       cpu: 250m
     limits:
-      memory: 512Mi
+      memory: 8192Mi
```
Comment (harshavardhana) [inline examples/minioinstance-with-external-service.yaml]: We shouldn't use limits due to bugs in CFS implementation Linux kernel. https://github.com/kubernetes/kubernetes/issues/67577#issuecomment-516661131

**Comment by nitisht** [comment]
> Can you also fix TLS.md and update it with some instructions on how to use this feature ?

Done

**Comment by nitisht** [comment]
Ready for review @harshavardhana @aead PTAL

File: pkg/controller/cluster/csr.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+	var err error
+
+	glog.V(0).Infof("Generating private key")
+	privateKey, err = newPrivateKey(constants.DefaultEllipticCurve)
+	if err != nil {
+		glog.Errorf("Unexpected error during the ECDSA Key generation: %v", err)
+		return nil, nil, err
+	}
+
+	privKeyBytes, err := x509.MarshalPKCS8PrivateKey(privateKey)
```
Comment (aead) [inline pkg/controller/cluster/csr.go]: Error handling?

File: pkg/controller/cluster/csr.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+func newPrivateKey(curve elliptic.Curve) (*ecdsa.PrivateKey, error) {
+	return ecdsa.GenerateKey(curve, rand.Reader)
+}
+
+func generateCryptoData(mi *miniov1beta1.MinIOInstance, serviceName string) ([]byte, []byte, error) {
+	var privateKey *ecdsa.PrivateKey
+	var err error
+
+	glog.V(0).Infof("Generating private key")
+	privateKey, err = newPrivateKey(constants.DefaultEllipticCurve)
```
Comment (aead) [inline pkg/controller/cluster/csr.go]:
```suggestion
	privateKey, err := newPrivateKey(constants.DefaultEllipticCurve)
```

Comment (aead) [reply pkg/controller/cluster/csr.go]: And delete L 56/57.

File: pkg/controller/cluster/csr.go (hunk 5 above)
Comment (nitisht) [reply pkg/controller/cluster/csr.go]: fixed

**Comment by nitisht** [comment]
PTAL @harshavardhana @aead

**Pull request merged**